File: 1-TwoSum/1-TwoSum.cpp
// Last updated: 6/3/2026, 11:35:14 PM
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) 
    {
        map<int, int> mpp;

        for (int i=0; i<nums.size(); i++){
            int left = target - nums[i]; 

            if (mpp.find(left) != mpp.end())
                return {mpp[left], i}; 
            
            mpp[nums[i]] = i; 
        }    

        return {}; 
    }
};

File: 1-TwoSum/1-TwoSum.py
# Last updated: 6/3/2026, 11:35:16 PM
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        """
        Returns indices of the two numbers in 'nums' that add up to 'target'.
        Exactly one valid solution exists in the input.
        """
        seen = {}
        
        for i, num in enumerate(nums):
            complement = target - num
            
            # If we've already seen the complement, return the indices
            if complement in seen:
                return [seen[complement], i]
            
            # Store the index of the current number
            seen[num] = i
        
        # By problem statement, this line should never be reached
        # (since one solution is guaranteed to exist)
        return []
